src/Shared/TaskParameter.cs

```diff
@@ -891,6 +763,25 @@ public void CopyMetadataTo(ITaskItem destinationItem)
                 // between items, and need to know the source item where the metadata came from
                 string originalItemSpec = destinationItem.GetMetadata("OriginalItemSpec");
 
+#if !TASKHOST
+                if (_customEscapedMetadata != null && destinationItem is IMetadataContainer destinationItemAsMetadataContainer)
+                {
+                    // The destination implements IMetadataContainer so we can use the ImportMetadata bulk-set operation.
+                    IEnumerable<KeyValuePair<string, string>> metadataToImport = _customEscapedMetadata
+                        .Where(metadatum => string.IsNullOrEmpty(destinationItem.GetMetadata(metadatum.Key)));
+
+#if FEATURE_APPDOMAIN
+                    if (!AppDomain.CurrentDomain.IsDefaultAppDomain())
+                    {
+                        // Linq is not serializable so materialize the collection before making the call.
+                        metadataToImport = metadataToImport.ToList();
+                    }
+#endif
+
+                    destinationItemAsMetadataContainer.ImportMetadata(metadataToImport);
```

Comment:
Based on that profile, maybe worth joining the `OrignalItemSpec` value into the `IEnumerable` to avoid immediately copying a new `ImmutableDictionary` right after this block - haven't checked to see if `IEnumerable.Concat()` or spread operator works efficiently or ends up allocating a collection. Otherwise a local function + `yield return` would work